docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,358 @@
+# ADR 101: Data Companion Pull API
+
+## Changelog
+
+- 2023-02-28: Renumber from 084 to 101 (@thanethomson)
+- 2022-12-18: First draft (@thanethomson)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+Following from the discussion around the development of [ADR 100][adr-100], an
+alternative model is proposed here for offloading certain data from nodes to a
+"data companion". This alternative model inverts the control of the data
+offloading process, when compared to ADR 082, from the node to the data
+companion.
+
+Overall, this approach provides slightly weaker guarantees than that of ADR 082,
+but represents a simpler model to implement.
+
+## Alternative Approaches
+
+Other considered alternatives to this ADR are also outlined in
+[ADR-100][adr-100].
+
+## Decision
+
+> This section records the decision that was made.
+
+## Detailed Design
+
+### Requirements
+
+Similar requirements are proposed here as for [ADR-100][adr-100].
+
+1. A node _must_ support at most one data companion.
+
+2. All or part of the following data _must_ be obtainable by the companion, and
+   as close to real-time as possible:
+   1. Committed block data
+   2. `FinalizeBlockResponse` data, but only for committed blocks
+
+3. The companion _must_ be able to establish the earliest height for which the
+   node has all of the requisite data.
+
+4. The API _must_ be (or be able to be) appropriately shielded from untrusted
+   consumers and abuse. Critical control facets of the API (e.g. those that
+   influence the node's pruning mechanisms) _must_ be implemented in such a way
+   as to eliminate the possibility of accidentally exposing those endpoints to
+   the public internet unprotected.
+
+5. The node _must_ know, by way of signals from the companion, which heights'
+   associated data are safe to automatically prune.
+
+6. The companion _must_ be able to handle the possibility that a node might
+   start from a non-zero height (i.e. that the node may state sync from a
+   specific height beyond genesis).
+
+7. The companion _must_ be able to handle the possibility that a node may
+   disable and then later re-enable the companion interface, potentially causing
+   the companion to have missing data in between those two heights.
+
+8. The API _must_ be opt-in. When off or not in use, it _should_ have no impact
+   on system performance.
+
+9. It _must_ not cause back-pressure into consensus.
+
+10. It _must_ not cause unbounded memory growth.
```

Comment:
It is not possible for something to meet the requirement that it  "_must_ ...  [provide] Committed block data" while also meeting the requirement that it "_must_ not cause back-pressure into consensus". These are mutually exclusive.

Reply:
I'm interested to know: how do you see these two aspects as mutually exclusive?

Reply:
If CometBFT wants to assume that sending messages to a configured data companion process is reliable, then it has to treat any message delivery error as fatal to the node. But even in that (definitely not viable in any real-world context) case, there's no way to guarantee that it will "not cause back-pressure to consensus", because message delivery success or failure is ultimately only known to the sender after a timeout, and that timeout necessarily impacts consensus.

Reply:
Are you perhaps thinking of the model proposed by ADR-100 in #73? Because the model in this ADR doesn't propose any requirements regarding guaranteed delivery of specific data to the companion; only that the node must facilitate some mechanism by which the companion can access that specific data should it need to. Requirement 2 is largely about which specific _kinds_ of data the node should make available to the companion.

In this model, it's up to the companion to establish whether or not it's reliably stored the data and manage the "retain height" that it indicates to the node accordingly (which influences the node's pruning mechanisms). Also in this model, data is "buffered" on disk, and not in memory. So inappropriate construction of a companion can result in unbounded _storage_ growth, yes, but that's not something within our control in this model.

Think of this model as an RPC interface for the companion to query whatever data it wants, whenever it wants, including the latest data. The companion also has a say in how the node prunes its data, and the node can only provide the data that it has (over which the companion has a fair degree of control).

Reply:
If a companion process is allowed to "lag" behind the block data of the node it is configured to pull from -- I agree that this is unavoidable, necessary, and ultimately not a problem -- then it doesn't satisfy requirement 1. Or, minimally, it means "as close to real time as possible" has no useful definition: it can be infinite.

In this case, the companion is simply a component to which the node sends state data, and proxies subsequent requests for that state data. But what then is the purpose? Whatever work the companion does, writing data to disk or whatever, could equally be done by the node itself, right? This physical separation of concerns has to be motivated by some requirement that is best satisfied by a separate process.

Reply:
> If a companion process is allowed to "lag" behind the block data of the node it is configured to pull from

This is precisely the difference between the two competing proposals of ADR-100 and ADR-101. I'll see if I can capture this more clearly in the ADR context.

> Or, minimally, it means "as close to real time as possible" has no useful definition: it can be infinite.

How would you articulate it better? The goal is to offer the data to the companion as quickly as reasonably possible after it's committed, with minimal interference with the node's core operations. I think it's important to keep some pragmatic perspective here on where we _are_ right now (see below), and it's pretty clear that infinite delivery time here would be useless. It's also clear that there's no way to _guarantee_ real-time delivery without the possibility of halting the node (which is the whole premise of ADR-100).

> But what then is the purpose?

Was the [context](https://github.com/cometbft/cometbft/blob/thane/adr-082-data-companion-api/docs/architecture/adr-100-data-companion-push-api.md#context) in ADR-100 not clear? We're ultimately trying to carve out a path here that:

1. Allows us to more clearly separate application- and consensus-related concerns
2. Doesn't immediately break the entire ecosystem

Currently, a node:

1. Handles a mixture of consensus- and application-related concerns, which results in massive storage bloat with application-specific data, and results in us having to maintain multiple indexers, complex query mechanisms, etc. (again, application-level concerns). This ultimately results in code bloat as we're forced to cater for more and more diverse application-specific use cases.
2. Has an incredibly flaky RPC that has a tendency to blow up the node the moment it incurs marginal load (and therefore is not scalable). This includes the _existing_ problem of slow WebSocket subscribers causing back-pressure into consensus. It's also not scalable for operators to have to run multiple full nodes in order to offer a robust RPC service to users, which is what some operators are forced to do, and it would, e.g., be preferable to scale the RPC independently of the node. But the best possible RPC here (especially the querying mechanisms) is application-specific.
3. Contains many layers of heavily intertwined technical debt, which significantly slows our progress in making meaningful changes to the system. We can't get rid of much of this right now due to (1) and (2).

We believe that we can solve many of these and related problems by more clearly separating application- and consensus-related concerns - even separating them into separate processes. We suspect that the first step is explicitly offering some way of offloading non-consensus-related functionality to a separate sidecar process (as captured in the two competing proposals of ADR-100 and ADR-101).

Once the necessary infrastructure is built out to support this functionality independently of the node, we will be in a better position to start to migrate more of the application-level concerns into the SDK/application, and we'll be able to further simplify the consensus engine.

Reply:
I definitely understand why you want to decouple consensus-related concerns from application-related concerns. But I don't understand how that is actually possible, because (AFAIU) the consensus layer is not invariant to the application layer. Or, equivalently: nothing prevents application-layer state from influencing consensus-layer decisionmaking. 

I could be mistaken! Am I mistaken? Are consensus-related concerns _guaranteed_ to be invariant to application-related concerns? If so, then I withdraw my objections 😉 But if not, then consensus is _necessarily_ coupled to the application.

Reply:
It's important to be concrete here about which specific application concerns bleed over into consensus where it's not necessary. The two areas articulated in these ADRs, specifically, are:

1. Events: https://github.com/cometbft/cometbft/blob/6088604212de8bed47c0c6cc7ebd7275069dcefc/proto/tendermint/abci/types.proto#L346
2. Transaction execution results: https://github.com/cometbft/cometbft/blob/6088604212de8bed47c0c6cc7ebd7275069dcefc/proto/tendermint/abci/types.proto#L351

This data is not necessary to facilitate consensus, and can only be properly interpreted by the application. We dump it to a database (in some cases taking up huge quantities of disk space) and offer it (rather poorly) via the RPC. This seems to have been incorporated into the application <-> consensus engine interaction model as a convenience for application developers, where all Tendermint Core-based nodes would offer a common RPC interface through which to query this data (e.g. `/block_search`, `/tx_search`, `/block_results`, the WebSocket subscription interface, etc.). This is one of those cases where, in trying to cater to everyone, we end up catering to no-one.

There are also more application concerns that have snuck into the consensus engine that really shouldn't be there now that we've got `PrepareProposal` and `ProcessProposal`, such as the notion of gas. But that's a whole other discussion (see [RFC 011](https://github.com/cometbft/cometbft/blob/main/docs/rfc/tendermint-core/rfc-011-delete-gas.md)).

One day in the distant future, the only thing that the consensus engine might be facilitating consensus on is an opaque set of bytes given to it by the application (which could represent transactions, an app hash, etc. - but should the consensus engine care about what those bytes mean?). Even the notion of "transactions" is effectively an application-level concern. But there's still a _lot_ of work to do to get there.

Reply:
Whether or not an application concern _should_ bleed into the consensus layer isn't really relevant. The question is if those application concerns _can possibly_ influence the consensus layer. If they can, then hiding application state behind an RPC means that state is not deterministic across nodes, which (AFAICT) breaks consensus.

Reply:
It's not clear to me that we're actually understanding each other here effectively. Feel free to [sign up](https://groups.google.com/g/cometbft-community/) to join one of our community calls to discuss this synchronously, or reach out to @adizere or @alijnmerchant21 if you'd like to have a more direct conversation!

Reply:
> For example, ABCI 1.0 defines a ProcessProposal method (which is part of consensus) that can reject proposed blocks. The implementation of that method is arbitrary

> But, otherwise, consensus-related concerns and application-related concerns are necessarily coupled, and (AFAICT) you can't evaluate one without knowledge of the other.

Did you bother to read the [ABCI 1.0 spec](https://github.com/cometbft/cometbft/blob/main/spec/abci/abci%2B%2B_app_requirements.md#formal-requirements)?

If the application is to repsect the ABCI 1.0 specification, it's implementation must definitely _not_  be arbitrary.